### PR TITLE
Vogue load bioinfo app part 1

### PR DESCRIPTION
**Pull request by hassanfa**
This PR adds vogue load bioinfo raw/process/sample to vogue app.

**How to prepare for test**:
- [x] ssh to hasta (depending on type of change)
- [x] install on stage

**How to test**:
Run cg upload vogue bioinfo and see it can read input options:
- [x] cg upload vogue bioinfo --case-name onemite --cleanup  --target-load all
- [x] cg upload vogue bioinfo --case-name onemite --cleanup  --target-load raw
- [x] cg upload vogue bioinfo --case-name onemite --cleanup  --target-load process

**Expected test outcome**:
- [x] no failed tests.
- [x] post screenshots of cli test explained above.

**Review:**
- [ ] code approved by 
- [x] tests executed by @hassanfa 
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Comment by hassanfa** [comment]
Pylint compain about `cg/cli/upload/vogue.py` should be ignored. As this PR is for setting up the framework and required tests.

Also, I don't know how to test an api. If anyone has any idea, please update the PR text body.

Otherwise, ready for review!

**Comment by hassanfa** [comment]
by testing an api, I mean testing on Hasta, screenshoting, etc... unittests are already added.

**Comment by hassanfa** [comment]
Fixing dummy_stdout text in tests...

**Comment by patrikgrenfeldt** [comment]
Oh I didn't follow, why can't this be fixed?
image

**Comment by patrikgrenfeldt** [comment]
I suggest in the test description something like when this is put in production how do you know it is working properly? And to put that as the test on stage, does ti make sense to you?

**Comment by hassanfa** [comment]
This PR is to create the API and framework. I like to work in increments.

Otherwise PR becomes too big. It is just an obvious and easy lint fix for the next PR.

**Comment by hassanfa** [comment]
> I suggest in the test description something like when this is put in production how do you know it is working properly? And to put that as the test on stage, does ti make sense to you?

I have explained above that I don't know how to test an internal API in production on hasta.

What do you think?

**Comment by patrikgrenfeldt** [comment]
> Also, I don't know how to test an api. If anyone has any idea, please update the PR text body.

There's a wikipage on the topic: https://en.wikipedia.org/wiki/API_testing :)

**Comment by hassanfa** [comment]
Cool! Unittests are covering them already. So it is ok. I'll update text body to say that.

Please merge it whenever then.

**Comment by hassanfa** [comment]
tests based on Patrick's criteria 👍 
image

**Comment by hassanfa** [comment]
@patrikgrenfeldt I hope these tests are this acceptable to you. 😁 It would be cool if you could lower bar a bit, so I can continue working on next parts of cg-vogue. 😉

File: cg/apps/vogue.py [hunk 1]
```diff
@@ -58,3 +58,65 @@ def load_flowcells(self, days):
...
+            load_bioinfo_process_inputs["case_analysis_type"],
+        ]
+
+        if cleanup_flag:
+            load_bioinfo_process_call.append("--cleanup")
+
+        self.process.run_command(load_bioinfo_process_call)
+
+    def load_bioinfo_sample(self, load_bioinfo_sample_inputs):
+        """Running load bioinfo sample."""
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply cg/apps/vogue.py:112]: I followed convention based on previous functions that do the same.

File: cg/apps/vogue.py [hunk 2]
```diff
@@ -58,3 +58,65 @@ def load_flowcells(self, days):
         # Execute command and print its stdout+stderr as it executes
         for line in self.process.stderr_lines():
             LOG.info("vogue output: %s", line)
+
+    def load_bioinfo_raw(self, load_bioinfo_raw_inputs):
+        """Running vogue load bioinfo raw."""
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply cg/apps/vogue.py:63]: I followed convention based on previous functions that do the same.

File: cg/apps/vogue.py [hunk 3]
```diff
@@ -58,3 +58,65 @@ def load_flowcells(self, days):
...
+            "--case-analysis-type",
+            load_bioinfo_raw_inputs["case_analysis_type"],
+            "--analysis-workflow",
+            load_bioinfo_raw_inputs["analysis_workflow_name"],
+        ]
+
+        self.process.run_command(load_bioinfo_raw_call)
+
+    def load_bioinfo_process(self, load_bioinfo_process_inputs, cleanup_flag):
+        """Running load bioinfo process."""
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply cg/apps/vogue.py:88]: I followed convention based on previous functions that do the same.

File: cg/cli/upload/vogue.py [hunk 4]
```diff
@@ -80,3 +71,32 @@ def samples(context, days: int):
...
+
+
+@vogue.command("bioinfo", short_help="Load bioinfo results to vogue")
+@click.option(
+    "-c",
+    "--case-name",
+    required=True,
+    help="Case name or project name for which the analysis results will load",
+)
+@click.option("--cleanup", default=False, help="Cleanup processed case data while loading")
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply cg/cli/upload/vogue.py]: it is based on Vogue's behavior. See new code. It is a proper flag now.

File: cg/cli/upload/vogue.py [hunk 5]
```diff
@@ -42,16 +37,12 @@ def genotype(context, days: int):
     context.obj["vogue_upload_api"].load_genotype(days=days)
 
 
-@vogue.command(
-    "apptags", short_help="Getting application tags to the trending database."
-)
+@vogue.command("apptags", short_help="Getting application tags to the trending database.")
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply cg/cli/upload/vogue.py:40]: I followed the text for previous code. not applicable to this PR.

File: cg/cli/upload/vogue.py [hunk 6]
```diff
@@ -80,3 +71,32 @@ def samples(context, days: int):
...
+@click.option("--cleanup", default=False, help="Cleanup processed case data while loading")
+@click.option(
+    "-t",
+    "--target-load",
+    default="all",
+    type=click.Choice(["all", "raw", "process"]),
+    show_default=True,
+    help=(
+        "Final target to load bioinfo data."
+        "Target orders are: all, raw, process."
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply cg/cli/upload/vogue.py:94]: See next four lines below where you commented.

File: cg/cli/upload/vogue.py [hunk 7]
```diff
@@ -80,3 +71,32 @@ def samples(context, days: int):
...
+        "Selecting all, will load raw, process, and sample."
+        "Selecting process, will upload raw and process."
+        "Selecting raw, will only load raw data."
+    ),
+)
+@click.pass_context
+def bioinfo(context, case_name, cleanup, target_load):
+    """Load bioinfo case results to the trending database"""
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply cg/cli/upload/vogue.py:102]: I followed convention based on previous commands that write the same.

File: tests/apps/vogue/test_vogue_api.py [hunk 8]
```diff
@@ -16,3 +19,135 @@ def test_instatiate(vogue_config):
...
+
+        # THEN assert that output config
+        assert load_bioinfo_sample_call_command in caplog.text
+
+
+def test_load_bioinfo_raw(vogue_config, caplog):
+
+    """Test load bioinfo raw upload in vogue api"""
+
+    # GIVEN a vogue api wit a config
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply tests/apps/vogue/test_vogue_api.py:59]: That's be too descriptive.

File: tests/apps/vogue/test_vogue_api.py [hunk 9]
```diff
@@ -16,3 +19,135 @@ def test_instatiate(vogue_config):
...
+
+    # WHEN instantiating a vogue api and input data
+    vogue_api = VogueAPI(vogue_config)
+    with mock.patch.object(subprocess, "run") as mocked:
+        mocked.return_value.stdout = dummy_stdout.encode("utf-8")
+        mocked.return_value.stderr = dummy_stderr.encode("utf-8")
+        mocked.return_value.returncode = dummy_returncode_success
+        vogue_api.load_bioinfo_raw(load_bioinfo_raw_inputs=test_load_bioinfo_raw_inputs)
+
+        # THEN assert that output config
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply tests/apps/vogue/test_vogue_api.py:103]: fxied

File: tests/apps/vogue/test_vogue_api.py [hunk 10]
```diff
@@ -16,3 +19,135 @@ def test_instatiate(vogue_config):
 
     # THEN assert that the adapter was properly instantiated
     assert vogue_api.vogue_binary == vogue_config["vogue"]["binary_path"]
+
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply tests/apps/vogue/test_vogue_api.py:22]: This is not my code. Previous PRs did this.

File: cg/cli/upload/vogue.py [hunk 11]
```diff
@@ -80,3 +71,32 @@ def samples(context, days: int):
...
+)
+@click.option("--cleanup", default=False, help="Cleanup processed case data while loading")
+@click.option(
+    "-t",
+    "--target-load",
+    default="all",
+    type=click.Choice(["all", "raw", "process"]),
+    show_default=True,
+    help=(
+        "Final target to load bioinfo data."
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply cg/cli/upload/vogue.py:93]: Please see Vogue help. irrelevant.

File: tests/apps/vogue/test_vogue_api.py [hunk 12]
```diff
@@ -16,3 +19,135 @@ def test_instatiate(vogue_config):
...
+        "load",
+        "bioinfo",
+        "sample",
+        "--analysis-case",
+        test_load_bioinfo_sample_inputs["analysis_case_name"],
+    ]
+    load_bioinfo_sample_call_command = " ".join(load_bioinfo_sample_call)
+    caplog.set_level(logging.INFO)
+
+    # WHEN instantiating a vogue api and input data
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply tests/apps/vogue/test_vogue_api.py:43]: one has to instantiate api to make input data work. irrelevant.

File: tests/apps/vogue/test_vogue_api.py [hunk 13]
```diff
@@ -16,3 +19,135 @@ def test_instatiate(vogue_config):
...
+        test_load_bioinfo_raw_inputs["analysis_workflow_version"],
+        "--case-analysis-type",
+        test_load_bioinfo_raw_inputs["case_analysis_type"],
+        "--analysis-workflow",
+        test_load_bioinfo_raw_inputs["analysis_workflow_name"],
+    ]
+    load_bioinfo_raw_call_command = " ".join(load_bioinfo_raw_call)
+    caplog.set_level(logging.INFO)
+
+    # WHEN instantiating a vogue api and input data
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply tests/apps/vogue/test_vogue_api.py:95]: Same as previous one. vogue api should be instantiated. irrelevant.

**Comment by hassanfa** [comment]
@patrikgrenfeldt addressed/commented on your comments.

**Comment by hassanfa** [comment]
Seem Pylint test is faulty, score is prefect 10/10:

image

**Comment by hassanfa** [comment]
Proof that Travis's Pylint is faulty: 
image

**Comment by patrikgrenfeldt** [comment]
> by testing an api, I mean testing on Hasta, screenshoting, etc... unittests are already added.

The PR description is "This PR adds vogue load bioinfo raw/process/sample to vogue app" so I guess that is what should be tested and screenshotted if possible, right?

**Comment by hassanfa** [comment]
> > by testing an api, I mean testing on Hasta, screenshoting, etc... unittests are already added.
> 
> The PR description is "This PR adds vogue load bioinfo raw/process/sample to vogue app" so I guess that is what should be tested and screenshotted if possible, right?

Could you scroll up a bit see the cli screenshot test I did? 😉 I also added description of tests. 

It is ok to reject the PR patrick, no need to drag it my friend.

**Comment by hassanfa** [comment]
I was planning to finalize Vogue parts today, it'd be super awesome if someone finalized this. I have already planned my next two weeks, and there is no Vogue in it.

File: tests/apps/vogue/test_vogue_api.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (patrikgrenfeldt) [reply tests/apps/vogue/test_vogue_api.py:43]: I don't think it is irrelevant. Being able to instantiate an API is an action that could and probably should be tested alone, calling a function is another action that probably should be tested alone. I'm suggesting that if you say something is a GIVEN like I think you did in this test then it is not also a part of the WHEN

**Comment by hassanfa** [comment]
@henrikstranneheim do you guys have a case in housekeeper that already has `multiqc.json` tag/file?

**Comment by henrikstranneheim** [comment]
This one should work:
```
housekeeper get -V -t multiqc-json onemite
301679 | /home/proj/stage/housekeeper-bundles/onemite/2020-02-06/multiqc_data.json | meta, multiqc-json, mip-dna, wgs, onemite
```

**Comment by henrikstranneheim** [comment]
Any MIP 8.0.0 production sample should have it

**Comment by hassanfa** [comment]
This is much larger than I anticipated. I don't know how to retrieve data for already analyzed samples from statusdb that are already in housekeeper and `context` in cg seems a bit more complex than thought.

Reason:
1. There are cases that some of their samples have empty `data_analysis` field.
2. There are samples their `delivery_at` is None but they have `data_analysis` field.
Example https://clinical-api.scilifelab.se/admin/familysample/?search=quickcorgi
3. There are cases that have tag in housekeeper but they don't have any workflow information on statusdb


The following information is needed to upload a case to vogue:
1. `case_id`: this is easy to get from status_db
2. `samples`: a comma separated list of samples for the case, but not all of them have data_analysis tag. How can I find out which sample has indeed been run and which one hasn't? `delivered_at` doesn't seem to be consistent. See point 1 and 2 above.
3. `multiqc.json`: This is easy to get through housekeeper API
4. `workflow_version`: Statusdb doesn't have this. Where can one find this value?

@patrikgrenfeldt could you tell me how can one find workflow version for a case that is stored in housekeeper? Also, how to find out which samples are run for the case and stored in housekeeper, is it statusdb or some other place?

**Comment by patrikgrenfeldt** [comment]
> @patrikgrenfeldt could you tell me how can one find workflow version for a case that is stored in housekeeper? Also, how to find out which samples are run for the case and stored in housekeeper, is it statusdb or some other place?

A case is currently a family entity in cg. Each analysed cases has a analyse entity connected to it that can be reached in an array in the family object as family.analyses. The latest analysis is family.analyses[0]. The analysis should contain information such as pipeline and pipelineversion. The data model of cg is located in cg/store/models.py.

The data-model has been expanded over time and some new fields like sample.data_analysis lack value for older cases.

**Comment by hassanfa** [comment]
@patrikgrenfeldt Any idea why Pylint is complaining about part of the code I haven't touched? It keeps failing? It says 
```
+cat /home/travis/pylint_after_output.txt
************* Module conftest
tests/meta/upload/vogue/conftest.py:46:0: R0903: Too few public methods (0/2) (too-few-public-methods)
tests/meta/upload/vogue/conftest.py:54:0: R0903: Too few public methods (1/2) (too-few-public-methods)
tests/meta/upload/vogue/conftest.py:71:0: R0903: Too few public methods (1/2) (too-few-public-methods)
```

This PR doesn't have touch that part of code, but still complains. Could you fix this?

**Comment by hassanfa** [comment]
Talk to PG and lint score is actually increased, but ci checks don't calculate delta properly.

**Comment by moonso** [comment]
Hi, I suggest we merge this as it is tomorrow after @hassanfa has tested it in stage. I'm working on a large PR for improving the test environment. When that is solved I will sit down with Hassan and write tests

**Comment by henrikstranneheim** [comment]
Sounds good to me. This is a bit of special circumstances and I trust you guys to complete the tests as soon as possible. I also think it will be much clearer (at least to me) how to write test in cg after Måns PR

**Comment by hassanfa** [comment]
Test 1, upload unprocessed `multiqc` data for `onemite` case, cli and mongodb stage:
1. cli:
image

2. Mongodb stage data:
image

**Comment by hassanfa** [comment]
Test 2 👍 , process an unprocessed `multiqc` on vogue database and populate bioinfo_process and bioinfo_sample collections. cli and mongodb test screenshots below:
1. cli:
image

2. 
image 
and
image

**Comment by hassanfa** [comment]
Test 3 👍 , a complete workflow of loading all the three collections in two previous tests:
cli and mongodb screenshots.

First removed all collections related to bioinfo for vogue-stage:
image

cli run to show it runs without problem:

1. image

2. three collections are populated: 
image

3. As an example, final sample data is populated on vogue-stage database:
image

**Comment by hassanfa** [comment]
Final ultimate über test that vogue-stage.scilifelab.se is showing data there: 👍 

image

**Comment by hassanfa** [comment]
Final ultimate über test that vogue-stage.scilifelab.se is showing data there: 👍

image

**Comment by patrikgrenfeldt** [comment]
Deployed on Hasta:
image

Deployed on Clinical-db: 
image

**Pull request merged**